golang/main.go

```diff
@@ -0,0 +1,183 @@
+		}
+		b.count = 0
+	}
+
+	return nil
+}
+
+func main() {
+	var kafkaURL string
+	if kafkaURL = os.Getenv("KAFKA_URL"); kafkaURL == "" {
+		kafkaURL = "localhost:9092"
+	}
+	var kafkaTopic string
+	if kafkaTopic = os.Getenv("EVENTS_TOPIC"); kafkaTopic == "" {
+		kafkaTopic = "test-topic"
+	}
+	seeds := []string{kafkaURL}
+	ctx := context.Background()
+
+	// Setup kafka consumer
+	cl, err := kgo.NewClient(
+		kgo.SeedBrokers(seeds...),
+		kgo.ConsumerGroup("default"),
+		kgo.ConsumeTopics(kafkaTopic),
+		kgo.DisableAutoCommit(),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer cl.Close()
+
+	var dbURL string
+	if dbURL = os.Getenv("DATABASE_URL"); dbURL == "" {
+		host := "localhost"
+		dockerized := os.Getenv("DOCKERIZED")
+		if dockerized != "" && dockerized != "0" && strings.ToLower(dockerized) != "false" {
+			host = "db"
+		}
+		if os.Getenv("POSTGRES_USER") == "" {
+			os.Setenv("POSTGRES_USER", "lotus")
+		}
+		if os.Getenv("POSTGRES_PASSWORD") == "" {
+			os.Setenv("POSTGRES_PASSWORD", "lotus")
+		}
+		if os.Getenv("POSTGRES_DB") == "" {
+			os.Setenv("POSTGRES_DB", "lotus")
+		}
+
+		dbURL = fmt.Sprintf("postgres://%s:%s@%s:5432/%s?sslmode=disable", os.Getenv("POSTGRES_USER"), os.Getenv("POSTGRES_PASSWORD"), host, os.Getenv("POSTGRES_DB"))
+	}
+	db, err := sql.Open("postgres", dbURL)
+	if err != nil {
+		panic(err)
+	}
+	defer db.Close()
+
+	insertStmt, err := db.Prepare("INSERT INTO metering_billing_usageevent (organization_id, customer_id, cust_id, event_name, time_created, properties, idempotency_id, inserted_at) VALUES ($1, $2, $3, $4, $5, $6, $7, $8) ON CONFLICT DO NOTHING")
+	if err != nil {
+		panic(err)
+	}
+	defer insertStmt.Close()
+
+	for {
+		fetches := cl.PollFetches(ctx)
+		if fetches == nil {
+			continue
+		}
+		if fetches.IsClientClosed() {
+			panic(errors.New("client is closed"))
+		}
+		if errs := fetches.Errors(); len(errs) > 0 {
+			// All errors are retried internally when fetching, but non-retriable errors are
+			// returned from polls so that users can notice and take action.
+			panic(fmt.Sprint(errs))
+		}
+
+		tx, err := db.Begin()
+		if err != nil {
+			fmt.Printf("Error starting transaction: %s\n", err)
+			continue
+		}
+		batch := &batch{
+			tx:         tx,
+			insertStmt: insertStmt,
+		}
+
+		fetches.EachRecord(func(r *kgo.Record) {
+			fmt.Printf("Received record: %s", r.Value)
+			var streamEvents StreamEvents
+			err := json.Unmarshal(r.Value, &streamEvents)
+			if err != nil {
+				fmt.Printf("Error unmarshalling event: %s\n", err)
+				return
+			}
+
+			if streamEvents.Event == nil {
+				if len(streamEvents.Events) > 0 {
+					streamEvents.Event = &streamEvents.Events[0]
+				} else {
+					fmt.Println("Error: both event and events fields are missing from stream_events")
+					return
+				}
+			}
+
+			event := streamEvents.Event
+			event.CustomerID = nil
+			event.InsertedAt = time.Now()
+
+			if err := batch.addRecord(event); err != nil {
+				fmt.Printf("Error inserting event: %s\n", err)
```

Comment:
Looking at the code for batch, I would definitely `panic` in this case. AFAICS the only thing that can go wrong in `batch` is either bugs in your code or a serious database failure/network partition of some kind. Because the usual referential integrity issues are already dealt with (`on conflict do nothing`), all that's left is bad stuff.

If you `panic` then the application goes back to a known state and can have another crack at it.